src/or/control.c

```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+  } else {
+    return 0;
+  }
+}
+
+/* Is the remote end of chan a client or bridge? */
+static int
+privcount_is_client(const channel_t* chan)
+{
+  int is_relay = privcount_is_consensus_relay(chan);
+  int is_client = privcount_is_unauthenticated_client(chan);
+
+  /* Relays in the consensus should always authenticate their peer IDs with us
+   */
+  if (is_relay) {
+    tor_assert_nonfatal(!is_client);
+  }
+
+  /* is_client is more reliable, but only with the fix for Tor bug 21406 */
+  return is_client && !is_relay;
+}
+
+/* Perform a 64-bit saturating add of a and b */
+static uint64_t
+privcount_add_saturating(uint64_t a, uint64_t b) {
+  if (PREDICT_LIKELY(UINT64_MAX - a > b)) {
+    return a + b;
+  } else {
+    return UINT64_MAX;
+  }
+}
+
+/* Perform a += b without overflow, and without returning a value */
+void
+privcount_sum(uint64_t *total, uint64_t increment) {
+  *total = privcount_add_saturating(*total, increment);
+}
+
+#define NO_CHANNEL_ADDRESS "0.0.0.0"
+#define NO_CONNECTION_ADDRESS "0.0.0.0"
+
+/* Return a newly allocated string containing the remote address of chan,
+ * or a placeholder if chan is NULL or has no connection.
+ * The returned string must be freed using tor_free(). */
+static char *
+privcount_chan_addr_to_str_dup(const channel_t *chan)
+{
+  /* channel_get_actual_remote_address / channel_tls_get_remote_descr_method
+   * can't be used here, because it uses a static buffer */
+  if (chan) {
+    tor_addr_t addr;
+    int has_addr;
+    /* channel_tls_get_remote_addr_method does not actually modify chan */
+    has_addr = channel_get_addr_if_possible((channel_t *)chan, &addr);
+    if (has_addr) {
+      return tor_addr_to_str_dup(&addr);
+    } else {
+      return tor_strdup(NO_CONNECTION_ADDRESS);
     }
+  } else {
+    return tor_strdup(NO_CHANNEL_ADDRESS);
+  }
+}
 
-    if(!conn || conn->base_.type != CONN_TYPE_EXIT) {
+/* Send a PrivCount DNS resolution event triggered on exitconn and orcirc.
+ * This event includes failed resolves, but excludes immediate results, such
+ * as trivial IP address resolves and failed malformed resolves.
+ * See PrivCount bug 184 for details. */
+void
+control_event_privcount_dns_resolved(const edge_connection_t *exitconn,
+                                     const or_circuit_t *orcirc)
+{
+    if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_DNS_RESOLVED)) {
+      return;
+    }
+
+    /* Filter out directory data (at the directory) and non-exit connections */
+    if (privcount_is_counted_for_dns(exitconn, orcirc)) {
         return;
     }
 
-    /* if the circuit started here, this is our own stream and we can ignore it */
-    circuit_t* circ = circuit_get_by_edge_conn(conn);
-    or_circuit_t *orcirc = NULL;
-    if(circ) {
-        if(CIRCUIT_IS_ORIGIN(circ)) {
-            return;
-        }
-        /* now we know its an or_circuit_t */
-        orcirc = TO_OR_CIRCUIT(circ);
+    /* There's no time here. We should fix that if we ever use this event */
```

Comment:
What does "There's no time here." mean? It isn't being included in the control event string? If so, it seems trivial to add it.